### PR TITLE
do not report 'noImplicitReturns' error if inferred return type of th…

### LINKED ISSUE #5814: noImplicitReturns compile error for void functions as of 1.8.0-dev.20151126
As of `1.8.0-dev.20151126` the following now generates a compile error when `noImplicitReturns` is specified in `tsconfig.json`:

```typescript
let x = () => { console.log("Test"); }
```

```
Not all code paths return a value.
```

even though the type of `x` is `() => void`

This feels like a bug to me...

### PR DESCRIPTION
**Pull request by vladima**
…e function is void/any. fixes #5814

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -9871,31 +9871,40 @@ namespace ts {
             }
 
             // Functions with with an explicitly specified 'void' or 'any' return type don't need any return expressions.
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: remove "explicitly specified"

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -9871,31 +9871,40 @@ namespace ts {
...
+            const hasExplicitReturn = func.flags & NodeFlags.HasExplicitReturn;
+
+            if (returnType && !hasExplicitReturn) {
+                // minimal check: function has syntactic return type annotation and no explicit return statements in the body 
+                // this function does not conform to the specification.
                 // NOTE: having returnType !== undefined is a precondition for entering this branch so func.type will always be present 
                 error(func.type, Diagnostics.A_function_whose_declared_type_is_neither_void_nor_any_must_return_a_value);
             }
+            else if (compilerOptions.noImplicitReturns) {
+                // errors in this branch should only be reported if CompilerOptions.noImplicitReturns flag is set
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: I think you can remove this comment

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -9871,31 +9871,40 @@ namespace ts {
...
+                    // Otherwise get inferred return type from function body and report error only if it is not void / anytype
+                    const inferredReturnType = hasExplicitReturn
+                        ? getReturnTypeOfSignature(getSignatureFromDeclaration(func))
+                        : voidType;
+
+                    if (inferredReturnType === voidType || isTypeAny(inferredReturnType)) {
+                        return;
+                    }
+                }
+                error(func.type || func, Diagnostics.Not_all_code_paths_return_a_value);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: What about `func.type || func.name || func`

File: src/compiler/checker.ts (hunk 1 above)
Comment (vladima) [inline src/compiler/checker.ts]: 'explicitly specified' reflects what is checked in the `if` below - if function has explicit type annotation and it is `void` / `any`

File: src/compiler/checker.ts (hunk 3 above)
Comment (vladima) [inline src/compiler/checker.ts]: `error` already picks `name` as a best error span for functions if this is possible

**Comment by vladima** [comment]
@DanielRosenwasser any other comments?

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -9871,31 +9871,40 @@ namespace ts {
...
-
-            // if return type is not specified then we'll do the check only if 'noImplicitReturns' option is set
-            if (!returnType && !compilerOptions.noImplicitReturns) {
+            if (returnType === voidType || isTypeAny(returnType)) {
                 return;
             }
 
             // If all we have is a function signature, or an arrow function with an expression body, then there is nothing to check.
-            // also if HasImplicitReturnValue flags is not set this means that all codepaths in function body end with return of throw
+            // also if HasImplicitReturn flag is not set this means that all codepaths in function body end with return of throw
```
Comment (sandersn) [inline src/compiler/checker.ts]: return **or** throw

**Comment by sandersn** [comment]
:+1:

File: src/compiler/checker.ts (hunk 1 above)
Comment (yuit) [inline src/compiler/checker.ts]: Remove the second "with"

**Comment by yuit** [comment]
:+1:

**Comment by DanielRosenwasser** [comment]
:+1:

**Pull request merged**